### PR TITLE
Migrate compile to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Migrate compile.dart and tests to null safety.

Part of #71511

### CODE REVIEW

**Comment by jmagman** [comment]
This one was tricky, would appreciate a thorough review...

File: packages/flutter_tools/lib/src/compile.dart [hunk 1]
```diff
@@ -216,26 +214,26 @@ class KernelCompiler {
   final List<String> _fileSystemRoots;
   final StdoutHandler _stdoutHandler;
 
-  Future<CompilerOutput> compile({
-    String sdkRoot,
-    String mainPath,
-    String outputFilePath,
-    String depFilePath,
+  Future<CompilerOutput?> compile({
+    required String sdkRoot,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/compile.dart:218]: This seems to actually be required, it's nullability isn't checked anywhere and it's always passed in.

File: packages/flutter_tools/lib/src/compile.dart [hunk 2]
```diff
@@ -314,7 +312,7 @@ class KernelCompiler {
         platformDill,
       ],
       ...?extraFrontEndOptions,
-      mainUri ?? mainPath,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/compile.dart:317]: `mainUri` isn't possible to be null as written:
```dart
String? mainUri;
if (packagesPath != null) {
  mainUri = packageConfig.toPackageUri(mainFileUri)?.toString(); // nullable
}
mainUri ??= toMultiRootPath(mainFileUri, _fileSystemScheme, _fileSystemRoots, _fileSystem.path.separator == r'\'); // nonnull
```
I'm not sure if this is a bug and there are cases when this should be `mainPath`.  But at least how it's written `mainUri` is always used in the command.

File: packages/flutter_tools/lib/src/compile.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart:218]: yes, definitely is

File: packages/flutter_tools/lib/src/compile.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart:317]: We added toMultiRootPath recently (relatively) so that probably changed things

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Comment by jmagman** [comment]
Google testing failing because artifacts nullability hasn't rolled.

**Pull request merged**